File: java21/src/main/java/io/github/sps23/interview/preparation/atomicity/TicketSnapshot.java
package io.github.sps23.interview.preparation.atomicity;

public record TicketSnapshot(int ticketsRemaining, boolean sellingOpen, String lastBuyer) {

    TicketSnapshot sellTo(String buyer) {
        var updatedRemaining = ticketsRemaining - 1;
        return new TicketSnapshot(updatedRemaining, updatedRemaining > 0, buyer);
    }
}

File: kotlin/src/main/kotlin/io/github/sps23/interview/preparation/atomicity/TicketSnapshot.kt
package io.github.sps23.interview.preparation.atomicity

data class TicketSnapshot(
    val ticketsRemaining: Int,
    val sellingOpen: Boolean,
    val lastBuyer: String?,
) {
    fun sellTo(buyer: String): TicketSnapshot {
        val updatedRemaining = ticketsRemaining - 1
        return TicketSnapshot(updatedRemaining, updatedRemaining > 0, buyer)
    }
}
